### DISCUSSION TITLE
Andrew Glassner: Deep Learning Crash Course

### DISCUSSION THREAD

**Discussion by kkmcgg** [NSCC-COGS/Aestheta#33] [Youtube Videos]
[![IMAGE ALT TEXT](https://img.youtube.com/vi/r0Ogt-q956I/0.jpg)](https://youtu.be/r0Ogt-q956I)

Honestly this is probably one of the best videos I've come across that balances covering a wide swath of deep learning tools while also going into a great bit of detail. Coincidentally - he specifically covers deep learning tools which we are likely to be interested in including  convolutional networks, autoencoders, and generative adversarial networks. Great Stuff!

**Comment by SabihornotSabih** [comment]
Thank you Kevin this is a great video! Lot to unpack from it. Really helpful in letting me piece together some stuff like the learning rate in gradient descent through backpropagation in neural nets.....interesting how Andrew said that it's not the error that gets moved in backpropagation but the weights, yet through backprop and the weights, one can use gradient descent to minimize the error or the loss. More subtle that way.
I'm interested in softmax would like to see more on that.
Had to go back and re-watch to better get regularization and drop-outs but the graphics help a lot. 
Convolution was also interesting but I think josh starmer explains it better also your explanations were also helpful. 
Most interesting for me was the autoencoder with the latent variables......sounded a lot like max pooling in convolution, compressing information down as much as possible into a few numbers... maybe it's not uncommon in machine learning...actually sounds like modern technology in general haha.

**Comment by SabihornotSabih** [comment]
Summary:
Machine learning can be thought of as a large class of algorithms in artificial intelligence studies, and deep learning is a particular style/method/approach/technique in machine learning. 
Deep learning can be supervised, unsupervised or reinforcement.  
Think of a neural net as an array of nodes, input, output or hidden nodes between the input and output nodes, arranged in layers (the nodes are 'neurons' in this analogy). 
Each hidden node is connected to all of the input nodes before it, and in each node there is an activation function that takes in input, each input attached to a weight which we control, and spits out an output. Several different types of functions are possible, for example, sigmoidal. 
The outputs are not probabilities which are what we need so we use a softmax function to turn the output scores into probabilities as numbers. 
The probabilities are then compared to the desired output labels and the difference or resulting loss is examined through a loss function. This loss or error is minimized using a technique called gradient descent (think of minimizing the slope or function in calculus). Gradient descent in turn is accomplished through another technique, or algorithm, called backpropagation, that manipulates or adjusts the weights (it works "backwards") to tell the gradient where to go in order to minimize it.  The steps that guide the gradient descent is through a number called the 'learning rate', where a bigger learning rate means bigger steps or jumps, while a smaller learning rate means smaller steps (especially needed when you get close to minimizing the error). 
A convolution neural net can be used to classify an image. It uses a filter (a small block of pixels) to overlay on the input image pixels and use backpropagation to determine the intensity of the filter. The filter is 'convolved' (involving dot product of the image and filter pixels) with the input image in order to best match with it. There can be many different filters needed, each with its own weight (for example each filter can be a particular color code). The filters spit out outputs that are mapped on a feature map that then go through an activation function (usually to ensure a binary range for example ReLu) before they are mapped onto another feature map. Max pooling is then used to compress the data of the feature map into the smallest feature map possible, the absolutely essential numbers. 
This concept is similar to the latent variables in autoencoding which is another different tool or technique, also used to compress data down and then make it big again (the latent variables are the absolutely reduced/compressed numbers that match the thousands of numbers used to describe the original image). 
A generative adversarial network is two networks running simultaneously: a generator that generates input image data, and a discriminator that compares the input to a desired output label and passes judgement  on whether or not it passes as identical or not, real or fake. We often then work with the discriminator to get better in not getting fooled, using backpropagation and gradient descent.

**Reply by kkmcgg** [reply]
I am very pleased that you took the time to go over this. I know its a lot of detail! A note to anyone else - if you do have the time I promise its worth while! If you watch this particular video a few times over you should begin to get a pretty firm base to work on in terms of how neural nets work. There have been more advancements that are kind of taking over the field in the last few years - but the general principals of back propagation and the linear algebra of forward-propagating network weights and biases still applies!  Personally I find autoencoders, GANS, and eigen images (he discusses eigen 'dogs') to be pretty well mind blowing concepts.
